app/controllers/auth/actions/AuthAction.scala

```diff
@@ -0,0 +1,210 @@
+            }
+          }
+        case Left(r)                     => Future.successful(r)
+      }
+    }
+  }
+
+  private def agentTokenCheck[A](request: Request[A], rq: => AuthenticatedRequest[A])(implicit
+    hc: HeaderCarrier
+  ): Future[Either[Result, AuthenticatedRequest[A]]] =
+    if (agentTokenCheck) {
+      dataCacheConnector.getAgentToken.map { agentToken =>
+        if (
+          (request
+            .getQueryString(Globals.TAXS_USER_TYPE_QUERY_PARAMETER)
+            .isEmpty || request
+            .getQueryString(Globals.TAXS_AGENT_TOKEN_ID)
+            .isEmpty) &&
+          agentToken.isEmpty
+        ) {
+          Left(notAuthorisedPage)
+        } else {
+          Right(rq)
+        }
+      }
+    } else {
+      Future.successful(Right(rq))
+    }
+
+  private def createAuthenticatedRequest[A](request: Request[A])(implicit
+    hc: HeaderCarrier
+  ): Future[Either[Result, AuthenticatedRequest[A]]] =
+    authorised(ConfidenceLevel.L50)
+      .retrieve(
+        Retrievals.allEnrolments and Retrievals.externalId and Retrievals.credentials and Retrievals.saUtr and Retrievals.nino and Retrievals.confidenceLevel
+      ) {
+        case Enrolments(enrolments) ~ Some(externalId) ~ Some(credentials) ~ saUtr ~ nino ~ confidenceLevel =>
+          val (agentRef, isAgentActive)           = agentInfo(enrolments)
+          def newRequest: AuthenticatedRequest[A] =
+            requests.AuthenticatedRequest(
+              userId = externalId,
+              agentRef = agentRef,
+              saUtr = saUtr.map(SaUtr),
+              nino = nino.map(Nino(_)),
+              isAgentActive = isAgentActive,
+              confidenceLevel = confidenceLevel,
+              credentials = credentials,
+              request = request
+            )
+          (agentRef.isDefined, isAgentActive) match {
+            case (true, false) =>
+              Future.successful(Left(Redirect(controllers.routes.ErrorController.notAuthorised)))
+            case (true, true)  =>
+              agentTokenCheck(request, newRequest)
+            case _             =>
+              pertaxAuthService.authorise[A, Request[A]](request).map {
+                case Some(r) => Left(r)
+                case _       => Right(newRequest)
+              }
+          }
+        case _                                                                                              => throw new RuntimeException("Can't find credentials for user")
+      } recover {
+      case _: NoActiveSession        =>
+        lazy val ggSignIn    = appConfig.loginUrl
+        lazy val callbackUrl = appConfig.loginCallback
+        Left(
+          Redirect(
+            ggSignIn,
+            Map(
+              "continue_url" -> Seq(callbackUrl),
+              "origin"       -> Seq(appConfig.appName)
+            )
+          )
+        )
+      case _: InsufficientEnrolments => Left(Redirect(controllers.routes.ErrorController.notAuthorised))
```

Comment:
Unreachable code. There is no restriction on enrolments in authorised(ConfidenceLevel.L50)

Reply:
Removed